### PR TITLE
Add tree to model

### LINKED ISSUE #23: hash all of the data fields in nodes
I think we need to hash all of the data fields in a FieldNode, regardless of whether or not we have encrypted data. This is because we want to be be able to ensure the integrity of the whole node, not just individual fields. The same goes for the other node types.

This will also help us verify that the redaction was actually performed correctly: i.e., if we have encrypted data, we should only have placeholder values in the plain data variables. If we don't hash all of the node variables together, we can't check that this property holds.

_Originally posted by @marcelamelara in https://github.com/TSELab/SBOMCtl/pull/20#discussion_r1826316817_

### PR DESCRIPTION
**Pull request by abuishgair**
Changes to complex node:
- Add type data field to distinguish between different types of complex nodes
- handle policy for complex type node
- added placeholders for node data fields and encrypted data.
Added document information as complex node such that the root only have complex nodes as children

Fixes #23

### CODE REVIEW

**Review by marcelamelara — CHANGES_REQUESTED** [review]
At a high level these changes LGTM @abuishgair ! But I would like to see #23 resolved as part of this PR as well, please.

File: privateSBOMExchange/src/petra/lib/models/__init__.py [hunk 1]
```diff
@@ -13,6 +13,8 @@
 
 from cpabe import cpabe_encrypt,cpabe_setup,cpabe_keygen,cpabe_decrypt
 
+data_placeholder:str="encrypted"
+encrypted_data_placeholder:str="public"
```
[the remark this answers is not in the document]
Comment (abuishgair) [reply privateSBOMExchange/src/petra/lib/models/__init__.py:17]: @marcelamelara do you mean uppercase it?

[the remark this answers is not in the document]
Comment (marcelamelara) [reply privateSBOMExchange/src/petra/lib/models/__init__.py:17]: Yes, constants in python are usually defined as all-caps variables.

[the remark this answers is not in the document]
Comment (abuishgair) [reply privateSBOMExchange/src/petra/lib/models/__init__.py:17]: @marcelamelara regarding 23, hashing all attributes , we have the hash and the policy as attributes does this mean removing them as attributes?

[the remark this answers is not in the document]
Comment (marcelamelara) [reply privateSBOMExchange/src/petra/lib/models/__init__.py:17]: What do you mean by "removing them as attributes"? Is there an issue with hashing the policy? And the hash variable in the node is needed to store the hash, no?

[the remark this answers is not in the document]
Comment (abuishgair) [reply privateSBOMExchange/src/petra/lib/models/__init__.py:17]: didn't fix 23 yet , @SantiagoTorres

File: privateSBOMExchange/src/petra/lib/models/__init__.py [hunk 2]
```diff
@@ -207,7 +207,7 @@ def visit_sbom_node(self, node:SbomNode):
             The computed hash of the SBOM node data and its children.
         """
         children_hashes = b''.join(child.accept(self) for child in node.children)
-        data_to_hash = (node.purl).encode() + children_hashes
+        data_to_hash = (f"{node.complex_type}{node.purl}").encode() + children_hashes
```
Comment (SantiagoTorres) [inline privateSBOMExchange/src/petra/lib/models/__init__.py:210]: Would we want to use a better encoding for the field information (e.g., as a json, or so?) I worry that (perhaps not for complex nodes) for some types we end up with fields that appear as other fields when we de-serialize.

**Review by SantiagoTorres — CHANGES_REQUESTED** [review]

File: privateSBOMExchange/src/petra/lib/models/__init__.py (hunk 2 above)
Comment (marcelamelara) [reply privateSBOMExchange/src/petra/lib/models/__init__.py:210]: I wouldn't want to hash json-encoded data, and I think it's actually a feature to obscure the nodes when the tree gets serialized. 

But you're right that we still want to be able to be able to reconstruct the tree correctly when we de-serialize. Here's the information per node we need to reconstruct: node type, encrypted_data, public_data, policy, hash, children/parents. So I can add a quick and dirty serialization/deserialization function to put trees on disk.

**Pull request merged**